source/sigv4.c

```diff
@@ -793,6 +854,236 @@ static SigV4Status_t getCredentialScope( SigV4Parameters_t * pSigV4Params,
     }
 
 /*-----------------------------------------------------------*/
+    static SigV4Status_t writeSignedHeaderToString( size_t headerIndex,
+                                                    uint32_t flags,
+                                                    CanonicalContext_t * canonicalRequest )
+    {
+        char * pBufLoc;
+        size_t buffRemaining, keyLen = 0, i = 0;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        pBufLoc = canonicalRequest->pBufCur;
+        buffRemaining = canonicalRequest->bufRemaining;
+
+        keyLen = canonicalRequest->pHeadersLoc[ headerNum ].key.dataLen;
+
+        if( keyLen + 1 > buffRemaining )
+        {
+            sigV4Status = SigV4InsufficientMemory;
+        }
+        else
+        {
+            for( i = 0; i < keyLen; i++ )
+            {
+                if( !( flags & SIGV4_HTTP_PATH_IS_CANONICAL_FLAG ) )
+                {
+                    *pBufLoc = tolower( canonicalRequest->pHeadersLoc[ headerNum ].key.pData[ i ] );
+                }
+                else
+                {
+                    *pBufLoc = canonicalRequest->pHeadersLoc[ headerNum ].key.pData[ i ];
+                }
+
+                pBufLoc++;
+            }
+
+            *pBufLoc = ';';
+            pBufLoc++;
+            buffRemaining -= ( keyLen + 1 );
+            canonicalRequest->pBufCur = pBufLoc;
+            canonicalRequest->bufRemaining = buffRemaining;
+        }
+
+        return sigV4Status;
+    }
+
+    static SigV4Status_t appendSignedHeaders( size_t headerCount,
+                                              uint32_t flags,
+                                              CanonicalContext_t * canonicalRequest )
+    {
+        size_t noOfHeaders = 0, keyLen = 0, i = 0;
+        SigV4Status_t sigV4Status = SigV4Success;
+
+        assert( canonicalRequest != NULL );
+        assert( canonicalRequest->pBufCur != NULL );
+
+        for( noOfHeaders = 0; noOfHeaders < headerCount; noOfHeaders++ )
+        {
+            assert( ( canonicalRequest->pHeadersLoc[ noOfHeaders ].key.pData ) != NULL );
+            sigV4Status = writeSignedHeaderToString( noOfHeaders, flags, canonicalRequest );
+
+            if( sigV4Status != SigV4Success )
+            {
+                break;
+            }
+        }
+
+        if( sigV4Status == SigV4Success )
+        {
+            canonicalRequest->pBufCur--;
+            *canonicalRequest->pBufCur = '\n';
+        }
+
+        return sigV4Status;
+    }
+
+    static SigV4Status_t writeCanonicalHeaderToString( size_t headerIndex,
```

Comment:
As this is generating the "CanonicalRequest", it should be named appropriately to not be confused with the string related to "String To Sign"
```suggestion
    static SigV4Status_t writeCanonicalHeaderToCanonicalRequest( size_t headerIndex,
```